### PR TITLE
llvm-doe: add new package

### PR DESCRIPTION
**Pull request by shintaro-iwasaki**
This PR creates a new package that provides a collection of the experimental LLVM projects created by the US DOE research and development teams.  Specifically, this package is expected to ease the ECP CI efforts for these projects.  This package contains:
- Clacc (https://csmd.ornl.gov/project/clacc)
- SOLLVE's loop transformation optimizations (https://www.bnl.gov/compsci/projects/SOLLVE/)
- SOLLVE BOLT OpenMP (https://www.bolt-omp.org/)

The default branch is `doe`, which is currently the same as the LLVM upstream but will merge all the efforts above.

The base of `package.py` is that of LLVM, but I dropped support of old versions.  The links are modified too.

All the components (`llvm-doe`, `llvm-doe@clacc`, `llvm-doe@pragma-omp-tile`,  ...) were successfully compiled and installed on our Intel Skylake server (Linux + GCC 7.5).

(@clementval, @jdoerfert, @Meinersbur)

### CODE REVIEW

**Comment by shintaro-iwasaki** [comment]
Could someone please review this PR?

File: var/spack/repos/builtin/packages/llvm-doe/package.py
Comment (alalazo) [inline var/spack/repos/builtin/packages/llvm-doe/package.py]: Where is this attribute used by `lmod`?

File: var/spack/repos/builtin/packages/llvm-doe/package.py
Comment (alalazo) [inline var/spack/repos/builtin/packages/llvm-doe/package.py]: Usually we name versions the same as the branches they refer to, but I know `master` etc. have special semantics for ordering. @adamjstewart What would be the way to go here?

File: var/spack/repos/builtin/packages/llvm-doe/package.py
Comment (alalazo) [inline var/spack/repos/builtin/packages/llvm-doe/package.py:38]: In general most of the code below seems duplicated from the `llvm` package. Would it be possible to coordinate with `llvm` maintainers and reduce code duplication to a minimum here? I think in this case it's particularly relevant, since the package recipe for `llvm` is fairly long and changes frequently.

File: var/spack/repos/builtin/packages/llvm-doe/package.py
Comment (shintaro-iwasaki) [reply var/spack/repos/builtin/packages/llvm-doe/package.py]: Thanks. `family = ` seems unnecessary as you pointed out.  I removed it.

File: var/spack/repos/builtin/packages/llvm-doe/package.py
Comment (shintaro-iwasaki) [reply var/spack/repos/builtin/packages/llvm-doe/package.py]: Thank you.  To avoid confusing and remove `preferred=True`, I changed `main` to `master` (and by doing so the `doe` branch becomes the default branch). `doe` is the branch name of the "default/main/master" branch of this repository.  If this naming is inappropriate, please let me know.

(As far as I checked `main` works as the default branch if `preferred=True` is set.  It seems that `main` has no meaning at this point.  I used `main` instead of `master` for a more inclusive branch name, but we should have followed the current naming rule since using `master` can avoid confusion.)

File: var/spack/repos/builtin/packages/llvm-doe/package.py
Comment (shintaro-iwasaki) [reply var/spack/repos/builtin/packages/llvm-doe/package.py:38]: I understand copying all is controversial, but we would like to maintain this package independently from the official LLVM package.

We'd like to do so because the upstream LLVM Spack package development is faster than that of this package.  If this package becomes a subclass of `Llvm`, for example, this package must be updated whenever the upstream LLVM Spack package is changed and breaks the compatibility.  As far as I tried previously, making a subclass of LLVM can surely reduce the duplication but it is fragile since this package class needs to depend on the internal variables and functions of LLVM. The code was not very clean since the LLVM package is not designed as a base class of another package (and I don't think the `llvm` maintainers want to redesign it).

A similar discussion happened in https://github.com/spack/spack/pull/12607, and at that time, the Spack developers allowed us to duplicate the code.

File: var/spack/repos/builtin/packages/llvm-doe/package.py
Comment (alalazo) [reply var/spack/repos/builtin/packages/llvm-doe/package.py]: To be clear, I don't mind calling the branch in any way that seems most appropriate to you as a maintainer. What I meant is that for versions tied to a branch most of the packages do:
```python
version('XYZ', branch='XYZ')
```
so I was just curious why the version was not `doe` in this particular case. Using `main`, `master`, `develop`, etc. triggers special semantics for comparison, as these names are treated like infinity and they compare greater as any other non-infinity version. See:

https://github.com/spack/spack/blob/de3504ca2965e8cdba4e615d6bcb590f274aebef/lib/spack/spack/version.py#L42-L43

and the associated Python module.

File: var/spack/repos/builtin/packages/llvm-doe/package.py
Comment (alalazo) [reply var/spack/repos/builtin/packages/llvm-doe/package.py:38]: > I understand copying all is controversial, but we would like to maintain this package independently from the official LLVM package.

In that case, do you need all the variants and choices from upstream or can you simplify this recipe?

File: var/spack/repos/builtin/packages/llvm-doe/package.py
Comment (shintaro-iwasaki) [reply var/spack/repos/builtin/packages/llvm-doe/package.py]: Oh I got it. Thank you. I changed it to the following so that the version name is the same as the branch name.
```py
    version('doe', branch='doe', preferred=True)
```

File: var/spack/repos/builtin/packages/llvm-doe/package.py
Comment (shintaro-iwasaki) [reply var/spack/repos/builtin/packages/llvm-doe/package.py:38]: We would like to keep all the variants so that one can install this as (at least at present) a replacement of `llvm` though I understand some configurations are rarely changed by users.

**Review by alalazo — APPROVED** [review]
I don't like much the code duplication here and in `sollve`, but we may work on reducing that in following PRs and rely on the maintainers of the package to keep the recipe as much as possible synced with upstream llvm.

@shintaro-iwasaki I'll give a couple of days for other people to chime in. If nobody has any comment ping me and I'll merge this PR.

**Comment by alalazo** [comment]
@shintaro-iwasaki As an aside, do you plan to remove the `sollve` package?

**Comment by shintaro-iwasaki** [comment]
@alalazo Thank you!

> As an aside, do you plan to remove the sollve package?

In my understanding, yes. Eventually `sollve` should be replaced by this new package, but I cannot guarantee when (e.g., right after this PR is merged) since some users including SOLLVE members might be using it.

**Comment by shintaro-iwasaki** [comment]
I asked the SOLLVE members and they said `sollve` can be removed once this patch is merged.  We will double-check if any user/CI is using it and if so check if it can be replaced by this new package.  We will remove `sollve` as soon as we confirm it, possibly within a week.

**Comment by adamjstewart** [comment]
See #20767 and #15896 for where I would like to go with package renames/removals. That is, once we have support for it, we should deprecate it for a full release before removing it entirely. Until #20767 is merged, it's "whatever goes".

**Comment by sethrj** [comment]
Is there a way to inherit from the parent `llvm` package and add additional variants? I have a strong aversion to code duplication.

**Comment by shintaro-iwasaki** [comment]
@sethrj I understand the concern of duplication.

As far as I tried previously, inheriting `llvm` was not only messy but also fragile because this package class needs to depend on the internal implementation of `llvm`.  Overall `llvm` is not designed to be inherited by another package, so using the parent `llvm` class is hard as far as I tried last time.  I don't think the `llvm` maintainers want to redesign and maintain `llvm` for this package.

We would be happy if this copy could be allowed. (Actually there are several changes to compile additional `llvm-doe` components so it is not a simple copy, and that's why it's hard to cleanly inherit `llvm`.)

**Comment by shintaro-iwasaki** [comment]
@alalazo We'd be happy if you could merge this PR.  After this PR, we will remove/deprecate `sollve` (maybe remove) soon.

**Comment by alalazo** [comment]
@shintaro-iwasaki Please ping me and I'll review the PR to remove `sollve`. Thanks!

**Comment by vlkale** [comment]
Has this PR been merged or will it be merged soon? It will be useful for the SOLLVE project and ECP efforts on CI for it to be so in the next few weeks if possible.

**Comment by alalazo** [comment]
@vlkale This was merged to `develop` 22 days ago :slightly_smiling_face:

**Comment by vlkale** [comment]
@alalazo Ah, oops. Just saw that now after looking again. Yeah, it’s an important package for SOLLVE project, so I just wanted to check. Thanks!

**Pull request merged**
